components/raftstore/src/store/fsm/peer.rs

```diff
@@ -5311,6 +5378,17 @@ where
             replicated_idx
         };
         assert!(compact_idx >= first_idx);
+
+        if let Some(idx) = witness_replicated_idx {
+            // 1. do not compact the log not replicated to witness, except witness is the
+            // most lagging peer
+            // 2. if there is a lagging behind follower, should not compact log, otherwise
+            // witness can't help the follower to catch up if the leader is down
+            if idx != replicated_idx {
+                compact_idx = std::cmp::min(compact_idx, replicated_idx);
```

Comment:
I'm afraid the problem solved by adjusting compact log index will all be brought back by witness.

Reply:
what's the problem

Reply:
Like OOM, log gc not in time, etc.

Reply:
It won't have more accumulated logs than before when witness is enabled. When the log gap reaches gc count max, PD would promote the witness to non-witness, then force gc can be performed.

Reply:
The process of generating snapshot, sending snapshot, applying snapshot can take minutes to finish in v1. I don't think "more accumulated logs than before" is true.